### DISCUSSION TITLE
Image | ROCK 5A

### DISCUSSION THREAD

**Discussion by krishnasen2401** [MichaIng/DietPi#6426] [Image Request]
#### Formal device information
- Device name | ROCK 5A
- Official product URL |  https://wiki.radxa.com/Rock5/5a
- GitHub resource URL (if available) | 
- Image download URL | 

#### Is the SBC officially supported by the Debian installer?
- ...

#### If not, is a reliable 3rd party Debian image available for this SBC?
https://github.com/radxa-build/rock-5a/releases/download/b16/rock-5a_debian_bullseye_kde_b16.img.xz

#### If not, are there install instructions for Debian available?
https://wiki.radxa.com/Rock5/linux/debian

**Comment by MichaIng** [comment]
Many thanks for your request.

Since neither Radxa nor Armbian do strictly differentiate between ROCK 5B and ROCK 5A (often named [ROCK 5](https://www.armbian.com/rock-5b/), did you try to boot with our ROCK 5B image?

**Reply by krishnasen2401** [reply]
I haven't tried it, but both have different processors 5b uses rk3588 and 5a uses rk3588s (same as rock 5, I think). Both have different official Debian images on the official website https://wiki.radxa.com/Rock5/downloads

**Reply by MichaIng** [reply]
There is no "ROCK 5", there is only "ROCK 5B" and "ROCK 5A": https://wiki.radxa.com/Rock5
But indeed Radxa provides two dedicated sets of images for them: https://wiki.radxa.com/Rock5/downloads

The ROCK 4 series all boot with the same image/bootloader, even that they also have different variants of the RK3399 SoC, so I was hoping it is the same with the ROCK 5 series. At least worth a test. The kernel btw provides a ROCK 5A device tree already, so only the bootloader is the relevant variable.

**Reply by krishnasen2401** [reply]
I will test the dietpi 5b build and check once

**Reply by krishnasen2401** [reply]
i am able to boot their official Debian but not the dietpi image of 5b

**Reply by MichaIng** [reply]
Okay, the latest U-Boot package from Armbian is quite old already (February), so we can retest once a new build is out. I'm not sure whether I will find time to otherwise generate own U-Boot builds soon. Here the repo and default config to start with, if someone wants to try: https://github.com/radxa/u-boot/blob/stable-5.10-rock5/configs/rock-5a-rk3588s_defconfig

**Reply by krishnasen2401** [reply]
Never had any experience in building Linux images, is there any guide I can follow

**Reply by Joulinar** [reply]
https://dietpi.com/docs/hardware/#make-your-own-distribution

**Reply by MichaIng** [reply]
Here it is about building U-Boot, not Linux and not an image 😉. We do it here for Quartz64: https://github.com/MichaIng/DietPi/blob/master/.github/workflows/quartz64.yml

What makes it complicated is that as Rockchip hardware is closed source, SPL binaries need to be downloaded and incorporated into the U-Boot build. Armbian btw compiles it from exactly the same source for all RK3588 boards: https://github.com/armbian/build/blob/main/config/sources/families/rockchip-rk3588.conf
But they use a different branch `next-dev` instead of `stable-5.10-rock5`. Ah, and now I see ROCK 5A is already WIP: https://github.com/armbian/build/blob/main/config/boards/rock-5a.wip
There you can see also the `defconfig`s used (including a dedicated one for SPI bootloader builds).

**Reply by krishnasen2401** [reply]
I have to flash the Uboot the SPI, right?
not sure if the kernel also needs to be changed

**Reply by MichaIng** [reply]
There is the following options:
- Boot via SPI bootloader from anywhere (SD card, eMMC, NVMe, USB). This generally works, but the bootloader that comes pre-flashed with the board seems to not support common `boot.scr` syntax and hence fails. There is an issue about this for ROCK 5B, which should apply as well to the ROCK 5A: #6431
- To fix the above, indeed flashing the SPI U-Boot build to the SPI flash would work and is probably the overall most convenient option, as then booting usually works with all images, regardless which bootloader they ship with (as it is not used).
- For us to provide ROCK 5A (compatible) images, we however want on on-disk bootloader that works, so that the image boots from SD card and eMMC, regardless which SPI bootloader the board is currently flashed with. So this would be the regular U-Boot build + flashing it to the SD card or eMMC or disk image file.

What I am not 100% sure about is in which order/priority which bootloader is used. In #6431 it seems that the bootloader from the disk image is used when there is one attached via SD card or eMMC, and the SPI bootloader is only used if no (bootable) SD card or eMMC is attached, so the SPI bootloader can further scan NVMe and USB. There is a maskrom button to disable the SPI bootloader and enforce booting from SD card/eMMC, not sure if there are other ways to set the boot priority or bootloader used. @cujomalainey I actually never asked whether when booting your ROCK 5B via SD card, is really the bootloader from SD card used, or is it actually still the same SPI bootloader? In theory the reason it does not boot our image via NVMe could be also related to some partitioning differences. Both images have GPT partition table and no dedicated boot partition, but at least the ROCK 5A image (I guess the ROCK 5B one as well) from Radxa have a little FAT partition at the beginning of the disk with some first boot config and a permanent `config.txt`, though empty. Who knows which subtile difference this makes to the boot process, I've seen a lot of hard to explain behaviours with vendor bootloaders.

**Reply by krishnasen2401** [reply]
@MichaIng is there a workflow for 3588 rock 5b?

**Reply by MichaIng** [reply]
As you can read by yourself, none found yet. I didn't ask: Did you try to boot via SD card of other storage media?

**Reply by krishnasen2401** [reply]
I am currently flashing the SPI my main system kept crashing while trying to flash the SD with the Debian image provided by the radxa . There is a Debian build instructions for Rock 5b, quite noticeable differences are Rock 5b has another m2 slot for SSDs and an HDMI input slot, so I am not sure if the 3588B kernel would work for both

**Reply by krishnasen2401** [reply]
it's not booting. I don't see anything on the screen

**Reply by cujomalainey** [reply]
@MichaIng yep SD card works fine for DietPi, its NVMe that is the issue. Did not try flash storage as I don't have any.

**Reply by cujomalainey** [reply]
Note there appears to be an issue with Radxa's spi bootloader, it says it will boot of SD card if available but that is definitely not true in practice. It definitely picks NVMe first. Even if I wiped the spi flash it will still boot the NVMe drive. Only solution i had was to nuke the NVMe drive with dd to get it back to the SD card.

**Reply by krishnasen2401** [reply]
> As you can read by yourself, none found yet. I didn't ask: Did you try to boot via SD card of other storage media?

I haven't tried m2 as this verison doesn't have m2 nvme slot need only M2 e key which is generally used for wifi need to get a converter, I haven't tried emmc as well don't have any on hand.

**Reply by krishnasen2401** [reply]
> Note there appears to be an issue with Radxa's spi bootloader, it says it will boot of SD card if available but that is definitely no true in practice. It definitely picks NVMe first. Even if I wiped the spi flash it will still boot the NVMe drive. Only solution i had was to nuke the NVMe drive with dd to get it back to the SD card.

I have instead on the SD card, nvme or emmc is not inserted at all, still it didn't boot or try to

**Reply by krishnasen2401** [reply]
One thing I noticed was I flashed uboot for emmc,SD,spi nor all three time it  booted the Debian from SD card

**Reply by krishnasen2401** [reply]
@MichaIng  Do you have the workflow or build script for 5B

**Reply by cujomalainey** [reply]
Figured out the issue. Radxa's bootload is very inconsistent with its bootloader order. If I flash one of their images it boots nvme consistently first. If I use dietpi it picks sdcard first. Use the armbian bootloader and shutdown, then pull the SD card bootup and it boots the NVMe fine.

**Reply by MichaIng** [reply]
Generally our image build workflow is:
- https://github.com/MichaIng/DietPi/blob/master/.github/workflows/dietpi-build.yml
- Invokes: https://github.com/MichaIng/DietPi/blob/master/.build/images/dietpi-build
- Invokes: https://github.com/MichaIng/DietPi/blob/master/.build/images/dietpi-installer
- And: https://github.com/MichaIng/DietPi/blob/master/.build/images/dietpi-imager

That boot order sounds a bit like how FriendlyELEC's bootloader does with SD card vs eMMC:
- If FriendlyELEC's bootloader is on eMMC, it prefers SD card
- If mainline bootloader is on eMMC, it prefers eMMC, which means: If there is a broken bootloader or image on eMMC, it can only be repaired via MASKROM button.

Don't ask me why they do it like that. A default that makes sense to me is preferring SD card whenever it contains any bootloader, then eMMC or NVMe, then USB, then network boot. At least having it consistent, regardless whether a mainline or vendor bootloader is present, and one removable with higher priority than any internal.

**Reply by cujomalainey** [reply]
> A default that makes sense to me is preferring SD card whenever it contains any bootloader, then eMMC or NVMe, then USB, then network boot.

Agreed, although I would expect USB before nvme/emmc but yes the order flipping is annoying. I wonder if it has to do with the type of script used to boot.

**Reply by MichaIng** [reply]
Armbian does have dedicated ROCK 5A images: https://dl.armbian.com/rock-5a/archive/
Only the bootloader build is different, but that one has not yet been uploaded to the APT repository. However, the ROCK 5B has all boot storage options (SPI, SD card, eMMC) of the ROCK 5A + M.2 NVMe socket, right? So I see no reason why the ROCK 5B bootloader would have any downside on ROCK 5A, as we know already that it does boot.

**Reply by MichaIng** [reply]
Armbian does have dedicated ROCK 5A images: https://dl.armbian.com/rock-5a/archive/
Only the bootloader build is different, but that one has not yet been uploaded to the APT repository. However, the ROCK 5B has all boot storage options (SPI, SD card, eMMC) of the ROCK 5A + M.2 NVMe socket, right? So I see no reason why the ROCK 5B bootloader would have any downside on ROCK 5A, as we know already that it does boot.

Should be fine to call it just "ROCK 5" without appendix on download page, banner etc, right? Or "ROCK 5 series" to clarify that all currently existing variants are supported? Or "ROCK 5A/5B" to be more concrete? 😄

**Reply by cujomalainey** [reply]
Yea that sounds correct. I would maybe just add a note to your docs pointing to a special case for NVMe to use the armbian bootloader and to make sure to pull the SD Card so future users don't get stuck.

**Reply by krishnasen2401** [reply]
The rock 5a has a m2 slot which a E key not popular with SSD or nvme ( I.e m or B key )

**Reply by MichaIng** [reply]
Ah I missed the still open topic with the SPI bootloader. @cujomalainey as the ROCK 5A has no NVMe capable M.2 slot, how exactly could it even try to boot from NVMe in your case with SPI bootloader?

@krishnasen2401 
If you have an Ethernet cable attached, so you see it in DHCP server/router UI and in case connect to the ROCK 5A via SSH? Would be theoretically possible that HDMI does not work. Otherwise, you do not have a serial adapter cable to check bootloader output, do you?

**Reply by cujomalainey** [reply]
@MichaIng my comment was specific to the 5B 😄

**Reply by MichaIng** [reply]
Okay, so we actually do not know whether our image boots on ROCK 5A then. Means I need to revert the name change for now.

**Reply by cujomalainey** [reply]
Sorry can't help ya there, only got the 5B on hand.

**Reply by MichaIng** [reply]
This request is about ROCK 5A explicitly 😉.

**Reply by cujomalainey** [reply]
> This request is about ROCK 5A explicitly 😉.

Yes but you looped me in from my other thread cause I was having boot issues as well on my 5B

**Comment by MichaIng** [comment]
Done: https://github.com/MichaIng/DietPi/commit/3a096eb

**Reply by krishnasen2401** [reply]
Will a new build be generated @MichaIng

**Reply by MichaIng** [reply]
Yes, the next days, changing the name of the image file and on download page. However, the image really is exactly the same, and the hardware model name shown in banner etc changes with the DietPi update. So you can use the ROCK 5B image now without any downsides.

**Reply by krishnasen2401** [reply]
I have to download the new dietpi 5b image and try again right?

**Reply by krishnasen2401** [reply]
Like the raspberry rock 5a has green blink and steady for boot success and red for no boot I will try again I did check with ethernet if possible post video for reference

**Reply by MichaIng** [reply]
I reverted the change. As cujomalainey was testing on ROCK 5B, quite possible that the image does not boot on ROCK 5A but that a different bootloader is required. In that case we'd need to wait for a package release of Armbian. Theoretically we could do own U-Boot builds, but I currently lack time for this.

**Reply by krishnasen2401** [reply]
I will re-download the latest build on Monday est and try it again, the last time I changes the spi flash and the Debian image booted everytime the flash instructions were for 5B and since no nvme but still it booted and I am not sure difference between a dietpi image and Debian script for dietpi convert, I can't try mask room as of now cause rock 5a doesn't have a switch per se like the 5b either I need male to male gpio cable to try cause there is no guide as of now.


I have uboot image but it's not currently on hand but it's the same instructions as dietpi5b just target is rock-5a instead of 5b I have check the terminal history

**Reply by krishnasen2401** [reply]
it is still the same the rock pi 5a is not booting up. I have checked the device on the networking using ethernet as well

**Reply by krishnasen2401** [reply]
@MichaIng there are build instruction at https://wiki.radxa.com/Rock5/guide/build-debian-from-debos-radxa but I am not sure how it related or dietpi or if there are similar at all

**Reply by MichaIng** [reply]
I'll implement it for next release, using kernel and bootloader from Armbian build system, which does now support the ROCK 5A.

**Comment by MichaIng** [comment]
A testing image is available here: https://dietpi.com/downloads/images/testing/
Would be great is someone could give is a try.

**Reply by krishnasen2401** [reply]
will give it a try

**Reply by krishnasen2401** [reply]
its booting up now

**Reply by MichaIng** [reply]
That is great. Let me know if anything is not working.

The MTD device for flashing an SPI bootloader is present, right?
```sh
ls -l /dev/mtdblock0
```

I.e. to enable USB boot, this should work:
```sh
source /usr/lib/u-boot/platform_install.sh
dd if="$DIR/rkspi_loader.img" of=/dev/mtdblock0 oflag=direct conv=notrunc status=progress
```

**Reply by MichaIng** [reply]
@krishnasen2401 
When you find time, can you check back regarding the SPI flash with above commands? So I can add an option for updating the SPI bootloader to `dietpi-config`.

**Comment by MichaIng** [comment]
All done, images can be found on our download page.
